api/rest/bundle_handler.go

```diff
@@ -0,0 +1,354 @@
+	}
+	if _, err := io.Copy(zipFile, rc); err != nil {
+		return fmt.Errorf("could not copy data to zip: %s", err)
+	}
+
+	return nil
+}
+
+func (h BundleHandler) Get(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	id := vars["uuid"]
+
+	bundle, err := h.getBundleState(id)
+	if err != nil {
+		writeJSONError(w, http.StatusNotFound, fmt.Errorf("bundle not found: %s", err))
+		return
+	}
+
+	body := jsonMarshal(bundle)
+	write(w, body)
+}
+
+func (h BundleHandler) GetFile(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	id := vars["uuid"]
+
+	http.ServeFile(w, r, filepath.Join(h.workDir, id, dataFileName))
+}
+
+func (h BundleHandler) List(w http.ResponseWriter, r *http.Request) {
+	ids, err := ioutil.ReadDir(h.workDir)
+	if err != nil {
+		writeJSONError(w, http.StatusInsufficientStorage, fmt.Errorf("could not read work dir: %s", err))
+	}
+
+	bundles := make([]Bundle, 0, len(ids))
+
+	for _, id := range ids {
+		if !id.IsDir() {
+			continue
+		}
+
+		bundle, err := h.getBundleState(id.Name())
+		if err != nil {
+			logrus.WithField("ID", id.Name()).WithError(err).Warn("There is a problem with bundle")
+		}
+		bundles = append(bundles, bundle)
+
+	}
+
+	body := jsonMarshal(bundles)
+
+	write(w, body)
+}
+
+//TODO(janisz): Add caching to this function
+func (h BundleHandler) getBundleState(id string) (Bundle, error) {
+	bundle := Bundle{
+		ID:     id,
+		Status: Unknown,
+	}
+
+	stateFilePath := filepath.Join(h.workDir, id, stateFileName)
+	rawState, err := ioutil.ReadFile(stateFilePath)
+	if err != nil {
+		return bundle, fmt.Errorf("could not read state file for bundle %s: %s", id, err)
+	}
+
+	err = json.Unmarshal(rawState, &bundle)
+	if err != nil {
+		return bundle, fmt.Errorf("could not unmarshal state file %s: %s", id, err)
+	}
+
+	if bundle.Status == Deleted || bundle.Status == Canceled || bundle.Status == Unknown {
+		return bundle, nil
+	}
+
+	dataFileStat, err := os.Stat(filepath.Join(h.workDir, id, dataFileName))
+	if err != nil {
+		bundle.Status = Unknown
+		return bundle, fmt.Errorf("could not stat data file %s: %s", id, err)
+	}
+
+	if bundle.Size != dataFileStat.Size() {
+		bundle.Size = dataFileStat.Size()
+		// Update status files
+		bundleStatus := jsonMarshal(bundle)
+		err = ioutil.WriteFile(stateFilePath, bundleStatus, filePerm)
+		if err != nil {
+			return bundle, fmt.Errorf("could not update state file %s: %s", id, err)
+		}
+	}
+
+	return bundle, nil
+}
+
+func (h BundleHandler) Delete(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	id := vars["uuid"]
+	stateFilePath := filepath.Join(h.workDir, id, stateFileName)
+
+	bundle, err := h.getBundleState(id)
+	if err != nil {
+		writeJSONError(w, http.StatusNotFound, fmt.Errorf("bundle not found: %s", err))
+		return
+	}
+
+	if bundle.Status == Deleted || bundle.Status == Canceled {
+		w.WriteHeader(http.StatusNotModified)
+		write(w, jsonMarshal(bundle))
+		return
+	}
+
+	//TODO(janisz): Handle Canceled Status
```

Comment:
Is that something we need to track in a JIRA ticket? Sounds pretty important to me for the functionality of this. What if we released without it? Also: Aren't we handling the `Canceled` state in line 301?

Reply:
We partially handle it there. We just do not cancel already cancelled job but there is no code for canceling job and removing created file.